File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/usecases/TransactionRejecting.java
package it.giannibombelli.wsc2026.payment.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.events.EventPublisher;
import it.giannibombelli.wsc2026.common.application.events.EventSubscriber;
import it.giannibombelli.wsc2026.payment.application.commands.RejectTransaction;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionRejected;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.application.policies.PaymentRejection;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;

import static java.util.Objects.requireNonNull;

public final class TransactionRejecting implements EventSubscriber<TransactionRejected> {
    private final PaymentRepository paymentRepository;
    private final EventPublisher<PaymentEvent> eventPublisher;
    private final PaymentRejection paymentRejection;

    public TransactionRejecting(PaymentRepository paymentRepository,
                                EventPublisher<PaymentEvent> eventPublisher,
                                PaymentRejection paymentRejection) {
        this.paymentRepository = requireNonNull(paymentRepository);
        this.eventPublisher = requireNonNull(eventPublisher);
        this.paymentRejection = requireNonNull(paymentRejection);
    }

    @Override
    public void on(TransactionRejected event) {
        Require.requireArgument(event, "event");

        RejectTransaction cmd = paymentRejection.evaluate(event);
        Payment payment = paymentRepository.findById(cmd.aggregateId())
            .orElseThrow(() -> new IllegalArgumentException("payment not found"));

        PaymentResultEvents.PaymentRejected rejected = payment.rejectTransaction(
            cmd.transactionId(),
            cmd.reason()
        );
        paymentRepository.save(payment);

        eventPublisher.publish(rejected);
    }
}

File: typescript/src/payment/application/usecases/transactionRejecting.ts
// Caso d'uso reattivo che rifiuta una transazione e marca il pagamento come REJECTED.

import type { EventPublisher } from '@/common/application/events/eventPublisher.js';
import type { EventSubscriber } from '@/common/application/events/eventSubscriber.js';
import { PaymentEvent } from '../../domain/events/paymentEvent.js';
import type { PaymentRejected } from '../../domain/events/paymentResultEvents.js';

import { TransactionRejected } from '../../domain/events/transactionRejected.js';
import { PaymentNotFoundException } from '../../domain/payment/paymentNotFoundException.js';
import { PaymentRejection } from '../../application/policies/paymentRejection.js';
import type { PaymentRepository } from '../../domain/ports/paymentRepository.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class TransactionRejecting implements EventSubscriber<TransactionRejected> {
  private readonly _paymentRepository: PaymentRepository;
  private readonly _eventPublisher: EventPublisher<PaymentEvent>;
  private readonly _paymentRejection: PaymentRejection;

  constructor(
    paymentRepository: PaymentRepository,
    eventPublisher: EventPublisher<PaymentEvent>,
    paymentRejection: PaymentRejection
  ) {
    requireDependency(paymentRepository, "paymentRepository");
    requireDependency(eventPublisher, "eventPublisher");
    requireDependency(paymentRejection, "paymentRejection");
    this._paymentRepository = paymentRepository;
    this._eventPublisher = eventPublisher;
    this._paymentRejection = paymentRejection;
  }

  on(event: TransactionRejected): void {
    requireArgument(event, 'event');

    const cmd = this._paymentRejection.evaluate(event);
    const payment = this._paymentRepository.findById(cmd.aggregateId);
    if (payment === null) throw new PaymentNotFoundException();

    const rejected: PaymentRejected = payment.rejectTransaction(cmd.transactionId, cmd.reason);
    this._paymentRepository.save(payment);

    this._eventPublisher.publish(rejected);
  }
}
